### PR TITLE
Support Batch Completion in Server

### LINKED ISSUE #2396: batching and streaming
Hi!
If i correctly understood code . Any type of entrypoints api with stream doesn't support batching . But LLMEngine has step() method for processing batch of requests and i can implement streaming with batch using step() or  there any pitfalls here? For example how handle these multiple response streams

### LINKED ISSUE #2441: API Server batch request issue
Hi, I am new to vLLM usage and i want to load and serve mistral 7b model using vLLM. Here is my brief understanding about vLLM. 
1. LLM Engine => could handle offline batching (i.e list of prompts)
2. Async LLM Engine => wrapped with LLM Engine and could server async calls individually but only through online serving (api_server.py)

Now I need to process batch calls (i..e list of prompts)  through API server batch request.
prompts = ["Give me a haiku poem"] * 10
How can I achieve this through api request. Kindly help me with this. Thanks in advance.

### PR DESCRIPTION
**Pull request by simon-mo**
OpenAI completion protocol support inputs for `prompt` to be `a string, array of strings, array of tokens, or array of token arrays.`. This PR adds support for that. 

Closes #2441 
Closes #2396

### CODE REVIEW

File: vllm/entrypoints/openai/serving_completion.py [hunk 1]
```diff
@@ -20,46 +21,55 @@
...
+                delta_text = res.prompt + output.text
+                delta_token_ids = res.prompt_token_ids + output.token_ids
+                top_logprobs = res.prompt_logprobs + (output.logprobs or [])
                 has_echoed[i] = True
+            else:
+                # return just the delta
+                delta_text = output.text[len(previous_texts[i]):]
+                delta_token_ids = output.token_ids[previous_num_tokens[i]:]
+                top_logprobs = output.logprobs[
+                    previous_num_tokens[i]:] if output.logprobs else None
```
Comment (simon-mo) [inline vllm/entrypoints/openai/serving_completion.py:60]: This part is refactoring for readability

File: vllm/entrypoints/openai/serving_completion.py [hunk 2]
```diff
@@ -129,51 +139,54 @@ def parse_prompt_format(prompt) -> tuple[bool, list]:
...
-            text=output_text,
-            logprobs=logprobs,
-            finish_reason=output.finish_reason,
-        )
-        choices.append(choice_data)
-
-    num_prompt_tokens = len(final_res.prompt_token_ids)
-    num_generated_tokens = sum(
-        len(output.token_ids) for output in final_res.outputs)
+            choices.append(choice_data)
```
Comment (simon-mo) [inline vllm/entrypoints/openai/serving_completion.py:184]: This part is refactoring for readability

**Review by zhuohan123 — APPROVED** [review]
LGTM! In the future, probably we can also add batched APIs to `AsyncLLMEngine`, so that we can reduce the number of async function calls.

**Comment by lixiaolx** [comment]
@simon-mo Hi, I am glad that the openai interface in vllm supports multiple batches, but I encountered the following two problems during actual testing:
1. First, when the openai version is high (for example, my local 0.28.0) https://community.openai.com/t/cannot-import-name-openai-from-openai/486147 , an error occurs and the interface inside needs to be recoded.
```
    from openai import OpenAI
ImportError: cannot import name 'OpenAI' from 'openai' (/usr/local/lib/python3.10/dist-packages/openai/__init__.py)

```

2. Second, when using echo=True, an error occurs
I positioned it here  https://github.com/vllm-project/vllm/blob/main/vllm/entrypoints/openai/serving_completion.py#L300 , can you fix it?
```
completion = openai.Completion.create(
    model=model,
    prompt="A robot may not injure a human being",
    echo=True,
    n=2,
    stream=stream,
    logprobs=3)

error:

vllm/entrypoints/openai/serving_completion.py", line 178, in request_output_to_completion_response
    output_text = prompt_text + output.text
TypeError: unsupported operand type(s) for +: 'NoneType' and 'str'
```

**Pull request merged**